### PR TITLE
Add [[nodiscard]] to cpp

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR adds [[nodiscard]] to many C++ API. It's largely an automatic update using clang-tidy -fix.

I disabled modernize-use-nodiscard for a single API: ice_getConnection on proxy, which is often used in our tests to establish a connection without actually retrieving the connection.

The addition of nodiscard also unveiled two bugs (in an IceGrid test and in IcePy). Both fixed by this PR.

### CODE REVIEW

**Review by pepone — APPROVED** [review]

File: cpp/include/DataStorm/InternalI.h [hunk 1]
```diff
@@ -51,7 +51,7 @@ namespace DataStormI
     {
     public:
         virtual ~KeyFactory() = default;
-        virtual std::shared_ptr<Key> get(std::int64_t) const = 0;
+        [[nodiscard]] virtual std::shared_ptr<Key> get(std::int64_t) const = 0;
         virtual std::shared_ptr<Key> decode(const Ice::CommunicatorPtr&, const Ice::ByteSeq&) = 0;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/include/DataStorm/InternalI.h:55]: clang-tidy only suggests (fixes) `[[nodiscard]]` for specific functions. See:
https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-nodiscard.html

We could add more `[[nodiscard]]` later on for additional functions, but this is for separate PRs not about lints.

File: cpp/src/DataStorm/ConnectionManager.cpp [hunk 2]
```diff
@@ -63,7 +63,7 @@ ConnectionManager::remove(const Ice::ConnectionPtr& connection)
     exception_ptr ex;
     try
     {
-        connection->getInfo();
+        [[maybe_unused]] auto _ = connection->getInfo();
```
Comment (externl) [inline cpp/src/DataStorm/ConnectionManager.cpp:66]: I'm not following this code. Doesn't `getInfo` only throw if the connection is closed. Are we relying on this behavior to get the close exception?

**Review by externl — APPROVED** [review]
Looks good. I fine the API usage of `getInfo` not great.

File: cpp/include/Ice/Service.h [hunk 3]
```diff
@@ -131,22 +131,22 @@ namespace Ice
...
-        std::string name() const;
+        [[nodiscard]] std::string name() const;
 
         /**
          * Determines whether the operating system supports running the
          * program as a Win32 service or Unix daemon.
          * @return True if the system supports services, false otherwise.
          */
-        bool checkSystem() const;
+        [[nodiscard]] bool checkSystem() const;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/include/Ice/Service.h]: Removed.

File: cpp/src/DataStorm/ConnectionManager.cpp (hunk 2 above)
Comment (bernardnormier) [reply cpp/src/DataStorm/ConnectionManager.cpp:66]: That's apparently what we do, yes. However, I am not following this code either.

**Pull request merged**